### ISSUE TITLE
make async nbdserver tests 100% reliable

### ISSUE THREAD

**Issue by GlenDC** [g8os/blockstor#80]
Right now the nbdserver tests which rely on testing content, that's modified/added/deleted asynchronously tend to fail (especially when the workload on unit testing increases).

Until now the way to test this was pretty brute force, by using the brute `time.Sleep`, but clearly that's not reliable enough, unless we want to keep it increasing...

This needs to be resolved, as it gives false CI failures.

**Comment by GlenDC** [comment]
Was thinking about this. So far I can only come up with the following things:

1. Redesign (part of) the code so that the async parts of code can be swapped out for sync parts;
2. Have a mocking version of the database server;
3. stick with the brute `time.Sleep` methods;

(2) could be a lot of work, with not too much gain. Although somehow this does sound like the most sane way to do testing of a system like this.

(1) is very practical, but we would lose some coverage. But in the end these are supposed to be unit tests, and not end-to-end tests. So perhaps this would allow us to separate the testing. By redesigning the code a bit we would with the current test, simply test the logic as a whole. And than we could add other tests that do keep the async parts in there, and they test if the code holds up in busy parallel environments.

(3) is the most practical, and in a way it should work as long as we keep the value high enough. On my dev machine the timers can be pretty low, but the travis environments seem to require a time higher than 100ms, for now I put the sleep timers at 200ms, and that seems to hold up for now. But this is really a flawed way, and it somehow exposes flaws in our design I think.

**Comment by GlenDC** [comment]
When we look at the travis logs where the build fails, and we compare that with the code and its flow, we do need to conclude that it always seems to fail because of the time ledisdb requires, and the fact that we don't wait for it's response.

Since https://github.com/g8os/blockstor/commit/3588ab4e349a5318e024b472362fd1e52b384b81 we don't wait for a response of the redis server, in the cases where we don't do anything with the response. Instead we simply `Send` the request and simply `Flush` it afterwards, and not receiving the response. Check [ardb/storage.go#L21-L31](https://github.com/g8os/blockstor/blob/3588ab4e349a5318e024b472362fd1e52b384b81/nbdserver/ardb/storage.go#L21-L31) to see how this is done.

The reason why this was done is because it gave a small speed gain in the code, but really the speed we gained because of it was/is marginal...

I now start to think that this perhaps not a good idea, as it means that even in a synchronous process, we might now potentially get issues. Merging for example consist of first Getting the content and than Setting it. In which case we could have an issue like this:

```
---> SET A, 10 --------------------------------------> DONE (10)
------> MERGE A, 01 --------> GET A ------- receive 00 ---------> SET A+B --------> DONE(01)
```

In this example we will overwrite the previously SET content, during the MERGE process, even though the merging was called after the content was set. Than again even if we do use the `Do` redigo.Conn function to Set the content, and thus waiting for a reply from the server (confirming the operation is done), we would still get the same issue in case the Setting and Merging calls from the above examples would be doen from different threads/servers.

This could be prevented by using stuff like `WATCH` during the Merging process, so that we get an error if that key we just retrieved (GET) was modified, while we were merging. In which case we could restart the Merging process, so that we do merge with the latest content.

So the above example can be prevented for synchronous process by using `conn.Do` (this would by the way also resolve the need for the `time.Sleep` calls  in [ardb_test.go#L35-L177](https://github.com/g8os/blockstor/blob/d505f637b1c242e78da53f9cd29c10ac4fc2240a/nbdserver/ardb/ardb_test.go#L35-L177)), and using `WATCH` during merging, so that also the asynchronous processes would handle that fine.

But this is just one random example, who knows what other flaws are there. I think it would be a good idea to add a series of tests for this.

**Comment by GlenDC** [comment]
If we remove the `redisSendNow` custom function, and instead use the `conn.Do` function as we did before, a lot of the `time.Sleep` functions are no longer required in the unit testing, and thus our code becomes a lot saner, as can be seen in the following diff:

```dif
diff --git a/nbdserver/ardb/ardb_test.go b/nbdserver/ardb/ardb_test.go
index 0c858c2..a9684da 100644
--- a/nbdserver/ardb/ardb_test.go
+++ b/nbdserver/ardb/ardb_test.go
@@ -4,7 +4,6 @@ import (
 	"bytes"
 	"errors"
 	"testing"
-	"time"
 
 	"github.com/g8os/blockstor/redisstub"
 	"github.com/garyburd/redigo/redis"
@@ -69,8 +68,6 @@ func testBackendStorage(t *testing.T, storage backendStorage) {
 			t.Fatal(err)
 		}
 	}
-	// give any async processes time to process
-	time.Sleep(time.Millisecond * 200)
 
 	// getting this content should now be possible
 	content, err = storage.Get(testBlockIndexA)
@@ -86,8 +83,6 @@ func testBackendStorage(t *testing.T, storage backendStorage) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	// give any async processes time to process
-	time.Sleep(time.Millisecond * 200)
 	// getting the content should now fail
 	content, err = storage.Get(testBlockIndexB)
 	if err != nil {
@@ -102,8 +97,6 @@ func testBackendStorage(t *testing.T, storage backendStorage) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	// give any async processes time to process
-	time.Sleep(time.Millisecond * 200)
 	// getting the content should be fine
 	content, err = storage.Get(testBlockIndexB)
 	if err != nil {
@@ -118,8 +111,6 @@ func testBackendStorage(t *testing.T, storage backendStorage) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	// give any async processes time to process
-	time.Sleep(time.Millisecond * 200)
 	// getting the content should be fine
 	content, err = storage.Get(testBlockIndexB)
 	if err != nil {
@@ -134,8 +125,7 @@ func testBackendStorage(t *testing.T, storage backendStorage) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	// give any async processes time to process
-	time.Sleep(time.Millisecond * 200)
+
 	// content should now be nil
 	content, err = storage.Get(testBlockIndexA)
 	if err != nil {
@@ -148,8 +138,7 @@ func testBackendStorage(t *testing.T, storage backendStorage) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	// give any async processes time to process
-	time.Sleep(time.Millisecond * 200)
+
 	// content should be merged
 	content, err = storage.Get(testBlockIndexA)
 	if err != nil {
@@ -164,8 +153,7 @@ func testBackendStorage(t *testing.T, storage backendStorage) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	// give any async processes time to process
-	time.Sleep(time.Millisecond * 200)
+
 	// content should now be nil
 	content, err = storage.Get(testBlockIndexB)
 	if err != nil {
diff --git a/nbdserver/ardb/deduped.go b/nbdserver/ardb/deduped.go
index 201e765..054ad1c 100644
--- a/nbdserver/ardb/deduped.go
+++ b/nbdserver/ardb/deduped.go
@@ -254,7 +254,7 @@ func (ds *dedupedStorage) getContent(hash blockstor.Hash) (content []byte, err e
 				return
 			}
 
-			err = redisSendNow(conn, "SET", hash.Bytes(), content)
+			_, err = conn.Do("SET", hash.Bytes(), content)
 			if err != nil {
 				// we won't return error however, but just log it
 				log.Infof("couldn't store remote content in local storage: %s", err.Error())
@@ -282,7 +282,7 @@ func (ds *dedupedStorage) setContent(prevHash, curHash blockstor.Hash, content [
 
 		exists, err := redis.Bool(conn.Do("EXISTS", curHash.Bytes()))
 		if err == nil && !exists {
-			err = redisSendNow(conn, "SET", curHash.Bytes(), content)
+			_, err = conn.Do("SET", curHash.Bytes(), content)
 		}
 
 		return
@@ -320,7 +320,7 @@ func (ds *dedupedStorage) goReferenceContent(hash blockstor.Hash) (err error) {
 	}
 	defer conn.Close()
 
-	err = redisSendNow(conn, "INCR", key)
+	_, err = conn.Do("INCR", key)
 	return
 }
 
@@ -348,7 +348,7 @@ func (ds *dedupedStorage) goDereferenceContent(hash blockstor.Hash) (err error)
 		return
 	}
 
-	err = redisSendNow(conn, "DEL", hash.Bytes(), key)
+	_, err = conn.Do("DEL", hash.Bytes(), key)
 	return
 }
 
diff --git a/nbdserver/ardb/deduped_test.go b/nbdserver/ardb/deduped_test.go
index 3054771..e53fdf9 100644
--- a/nbdserver/ardb/deduped_test.go
+++ b/nbdserver/ardb/deduped_test.go
@@ -72,7 +72,6 @@ func createTestDedupedStorage(t *testing.T, vdiskID string, blockSize, blockCoun
 // testDedupContentExists tests if
 // the given content exists in the database
 func testDedupContentExists(t *testing.T, memRedis *redisstub.MemoryRedis, content []byte) {
-	time.Sleep(time.Millisecond * 200) // give background thread time
 	conn, err := memRedis.Dial("")
 	if err != nil {
 		debug.PrintStack()
@@ -98,7 +97,6 @@ func testDedupContentExists(t *testing.T, memRedis *redisstub.MemoryRedis, conte
 // testDedupContentDoesNotExist tests if
 // the given content does not exist in the database
 func testDedupContentDoesNotExist(t *testing.T, memRedis *redisstub.MemoryRedis, content []byte) {
-	time.Sleep(time.Millisecond * 200) // give background thread time
 	conn, err := memRedis.Dial("")
 	if err != nil {
 		debug.PrintStack()
@@ -279,8 +277,6 @@ func TestGetDedupedRootContent(t *testing.T) {
 	if content == nil {
 		t.Fatal("content should exist now, while received nil-content")
 	}
-	// give database some time to process the async storing
-	time.Sleep(time.Millisecond * 200)
 
 	// content should now be in both storages
 	// as the remote get should have also stored the content locally
@@ -315,8 +311,6 @@ func TestGetDedupedRootContent(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	// give database some time to process the flush
-	time.Sleep(time.Millisecond * 200)
 
 	// let's copy the metadata from storageB to storageA
 	copyTestMetaData(t, vdiskIDB, vdiskIDA, redisProviderB, redisProviderA)
@@ -357,8 +351,6 @@ func TestGetDedupedRootContent(t *testing.T) {
 	if content == nil {
 		t.Fatal("content should exist now, while received nil-content")
 	}
-	// give database some time to process the async storing
-	time.Sleep(time.Millisecond * 200)
 
 	// and also our direct test should show that
 	// the content now exists in both storages
diff --git a/nbdserver/ardb/nondeduped.go b/nbdserver/ardb/nondeduped.go
index 75b3445..3abe6d8 100644
--- a/nbdserver/ardb/nondeduped.go
+++ b/nbdserver/ardb/nondeduped.go
@@ -38,12 +38,12 @@ func (ss *nonDedupedStorage) Set(blockIndex int64, content []byte) (err error) {
 	// don't store zero blocks,
 	// and delete existing ones if they already existed
 	if ss.isZeroContent(content) {
-		err = redisSendNow(conn, "DEL", key)
+		_, err = conn.Do("DEL", key)
 		return
 	}
 
 	// content is not zero, so let's (over)write it
-	err = redisSendNow(conn, "SET", key, content)
+	_, err = conn.Do("SET", key, content)
 	return
 }
 
@@ -70,7 +70,7 @@ func (ss *nonDedupedStorage) Merge(blockIndex, offset int64, content []byte) (er
 	copy(origContent[offset:], content)
 
 	// store new content, as the merged version is non-zero
-	err = redisSendNow(conn, "SET", key, origContent)
+	_, err = conn.Do("SET", key, origContent)
 	return
 }
 
diff --git a/nbdserver/ardb/storage.go b/nbdserver/ardb/storage.go
index 2028ded..0521b64 100644
--- a/nbdserver/ardb/storage.go
+++ b/nbdserver/ardb/storage.go
@@ -18,18 +18,6 @@ type backendStorage interface {
 	GoBackground(ctx context.Context)
 }
 
-// redisSendNow is a utitlity function used by backendStorage functions,
-// and is similar to redis.Conn.Do, except that we don't read the reply
-func redisSendNow(conn redis.Conn, cmd string, args ...interface{}) (err error) {
-	err = conn.Send(cmd, args...)
-	if err != nil {
-		return
-	}
-
-	err = conn.Flush()
-	return
-}
-
 // redisBytes is a utility function used by backendStorage functions,
 // where we don't want to trigger an error for non-existent (or null) content.
 func redisBytes(reply interface{}, replyErr error) (content []byte, err error) {
```

I think we should apply this patch, but I'm open to different opinions. This will by the way also make our travis CI builds be reliable once again, as waiting for goroutines seems pretty reliable. At least we didn't have issues before with it.

So while it does resolve this issue, and makes our code better, I think it would be good if we can remove all `time.Sleep` functions from our test code, as waiting for async threads using sleeps, smells.

So with the above patch applied, we still have `time.Sleep` in 2 situations (which only relate to the `dedupedStorage`):

1. `(*dedupedStorage).Flush`: This function is called when the nbdserver flushes. We call it in the unit tests, as we want to serialize the LBA shards into ledisdb, such that we can copy the metadata, as we do some tests related to that. I start to have a feeling that we can resolve this, by also waiting for a reply from the server, as that would allow the `(*dedupedStorage).Flush` to be exiting only when the server is done with the serialization, which might be desired?!
2. _Reference Counting_: The referencing and dereferencing (increasing/decreasing a counter, and deleting content when that counter reached 0), are done on a separate go-routine. In the unit tests we test that reference counter at each point, so we need to wait for the async process (uses channels internally) to finish. Channels and goroutines are pretty fast though, and very reliable, so time.Sleep for this in unit tests isn't too worry some imho, yet it does smell, so perhaps we could replace this part with an interface, so that we could make this part synchronously in the unit tests;

Open for suggestions though. My entire hope by writing this all out, is that anyone can give input on this topic.

**Comment by iwanbk** [comment]
```
Since 3588ab4 we don't wait for a response of the redis server, in the cases where we don't do anything with the response. Instead we simply Send the request and simply Flush it afterwards, and not receiving the response
```
Why we don't wait for the response?
It sounds to me like calling a `func` and doesn't wait for the response.

Sorry if it sounds as a silly question, i never give close attention to nbdserver before.

**Comment by GlenDC** [comment]
The theory was that we didn't use the reply in these values, but now that I think about, it also means we don't really see the error result, so yeah, it was a bad idea, and the speed gain was marginal anyhow.

**Comment by GlenDC** [comment]
Once PR #82 is merged into master, we have reliable Travis CI tests again. There are still a few `time.Sleep` calls in the `dedupedStorage` test code, which are used for 2 situations, both because of the internal use of go-routines:

+ When getting content, which is not available in the local storage, but is available in the remote storage, we store that content in a go-routine before returning the content to the caller. As a byproduct this means that when that function exists, we can't immediately test if that content is now available in local storage;
+ Testing reference counters, right after a dedupedStorage operation, requires a `time.Sleep`, as that is a process happening (via channels) on separate work goroutines;

With some design changes we could probably provide a separation of concerns, and make it possible to run that code synchronously in test code, while keeping it asynchronously in dev/prod. However I'm not convinced that this is really required. Goroutines are pretty reliable, and so I would be surprised if the smelly `time.Sleep` won't do the job forever, for those unit tests where we want to test these things.

But hey, I've been surprised before, and I'm sure things will keep surprising me in the future :)
